packages/embarkjs/ens/src/index.js

```diff
@@ -154,47 +156,74 @@ __embarkENS.registryAddresses = {
   "4": "0xe7410170f87102DF0055eB195163A03B7F2Bff4A"
 };
 
+function setProvider(web3, kind, endpoint) {
+  web3.setProvider(new Web3.providers[kind](endpoint));
+}
+
+function connectWebSocket(web3, endpoint, callback) {
+  setProvider(web3, 'WebsocketProvider', endpoint);
+  checkConnection(web3, callback);
+}
+
+function connectHttp(web3, endpoint, callback) {
+  setProvider(web3, 'HttpProvider', endpoint);
+  checkConnection(web3, callback);
+}
+
+function checkConnection(web3, callback) {
+  web3.eth.getAccounts(error => {
+    if (error) {
+      web3.setProvider(null);
+    }
+    callback(null, {
+      connected: !error,
+      error
+    });
+  });
+}
+
+__embarkENS.web3 = new Web3();
+
 __embarkENS.setProvider = function(config) {
   const self = this;
   const ERROR_MESSAGE = 'ENS is not available in this chain';
   self.registration = config.registration;
   self.env = config.env;
   self.ready = false;
 
-  // FIXME EmbarkJS.onReady doesn't work. Possibility of a race condition
-  EmbarkJS.onReady(err => {
-    if (err) {
-      throw new Error(err);
+  let connectionErrors = {};
+
+  reduce(config.dappConnection, false, (result, connectionString, next) => {
+    if (result.connected) {
+      return next(null, result);
     }
-    EmbarkJS.Blockchain.blockchainConnector.getNetworkId()
-      .then((id) => {
-        const registryAddress = self.registryAddresses[id] || config.registryAddress;
-        self._isAvailable = true;
-        self.ens = new EmbarkJS.Blockchain.Contract({
-          abi: config.registryAbi,
-          address: registryAddress,
-          web3: EmbarkJS.Blockchain.blockchainConnector.getInstance()
-        });
-        self.registrar = new EmbarkJS.Blockchain.Contract({
-          abi: config.registrarAbi,
-          address: config.registrarAddress,
-          web3: EmbarkJS.Blockchain.blockchainConnector.getInstance()
-        });
-        self.resolver = new EmbarkJS.Blockchain.Contract({
-          abi: config.resolverAbi,
-          address: config.resolverAddress,
-          web3: EmbarkJS.Blockchain.blockchainConnector.getInstance()
-        });
-        self.ready = true;
-      })
-      .catch(err => {
-        self.ready = true;
-        if (err.message.indexOf('Provider not set or invalid') > -1) {
-          console.warn(ERROR_MESSAGE);
-          return;
-        }
-        console.error(err);
-      });
+
+    if ((/^wss?:\/\//).test(connectionString)) {
+      connectWebSocket(self.web3, connectionString, next);
+    } else {
+      connectHttp(self.web3, connectionString, next);
+    }
+  }, async (err, result) => {
+    if (!result.connected || result.error) {
+      console.error(result.error);
+    }
+    const accounts = await self.web3.eth.getAccounts();
+    self.web3.eth.defaultAccount = accounts[0];
+    self.web3.eth.net.getId().then(id => {
```

Comment:
Why not use `await` here?

Reply:
Honestly, the reason for that was that I wanted to stay as close as possible to the previous version just to avoid possible errors unrelated to the actual functionality.

Changing this to async/await now :)